### PR TITLE
[GOBBLIN-1937] Quantify Missed Work Completed by Reminders

### PR DESCRIPTION
**Pull request by umustafi**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1937

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
Measure the amount of work that would have been missed by a participant otherwise completed by the reminder event concept. This can be done by checking the number of reminder events that actually obtain a lease to re-trigger a flow launch. Reminder events that see another lease is in progress are not included to avoid double counting the participant actually doing the work on reminder case. 
Additionally, address a bug to filter out heartbeat events before attempting to extract the FlowActionType from DagActionStoreChangeEvents because heartbeat events have null FlowActionTypes, causing the heartbeats to be rightfully skipped but for the wrong reason. 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
N/A metric emission 

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Review by Will-Lo — CHANGES_REQUESTED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/FlowTriggerHandler.java [hunk 1]
```diff
@@ -79,6 +79,7 @@ public class FlowTriggerHandler {
   private ContextAwareCounter noLongerLeasingStatusCount;
   private ContextAwareCounter jobDoesNotExistInSchedulerCount;
   private ContextAwareCounter failedToSetEventReminderCount;
+  private ContextAwareMeter numRemindersDoingMissedWork;
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/FlowTriggerHandler.java:82]: I also prefer the `num` prefix, but `count` made more sense for some of the names above so I'll stick with that convention.

File: gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/ServiceMetricNames.java [hunk 2]
```diff
@@ -35,13 +35,14 @@ public class ServiceMetricNames {
...
-  public static final String FLOW_TRIGGER_HANDLER_PREFIX = "flowTriggerHandler";
-  public static final String GOBBLIN_FLOW_TRIGGER_HANDLER_NUM_FLOWS_SUBMITTED = ServiceMetricNames.GOBBLIN_SERVICE_PREFIX + "." + FLOW_TRIGGER_HANDLER_PREFIX + ".numFlowsSubmitted";
-  public static final String FLOW_TRIGGER_HANDLER_LEASE_OBTAINED_COUNT = GOBBLIN_SERVICE_PREFIX + "." + FLOW_TRIGGER_HANDLER_PREFIX + ".leaseObtained";
-  public static final String FLOW_TRIGGER_HANDLER_LEASED_TO_ANOTHER_COUNT = GOBBLIN_SERVICE_PREFIX + "." + FLOW_TRIGGER_HANDLER_PREFIX + ".leasedToAnother";
-  public static final String FLOW_TRIGGER_HANDLER_NO_LONGER_LEASING_COUNT = GOBBLIN_SERVICE_PREFIX + "." + FLOW_TRIGGER_HANDLER_PREFIX + ".noLongerLeasing";
-  public static final String FLOW_TRIGGER_HANDLER_JOB_DOES_NOT_EXIST_COUNT = GOBBLIN_SERVICE_PREFIX + "." + FLOW_TRIGGER_HANDLER_PREFIX + ".jobDoesNotExistInScheduler";
-  public static final String FLOW_TRIGGER_HANDLER_FAILED_TO_SET_REMINDER_COUNT = GOBBLIN_SERVICE_PREFIX + "." + FLOW_TRIGGER_HANDLER_PREFIX + ".failedToSetReminderCount";
+  public static final String FLOW_TRIGGER_HANDLER_PREFIX = GOBBLIN_SERVICE_PREFIX + "." + "flowTriggerHandler";
```
[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/ServiceMetricNames.java:38]: updated the prefix and tried to change all the instances of it being used with `+ ".`

[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/ServiceMetricNames.java:38]: Actually I'm going to undo this for now bc I see we use it like this in metrics to create a name and in those cases an extra dot is added. I will instead create a different prefix with the dot and name it differently. 
`MetricRegistry.name(ServiceMetricNames.GOBBLIN_SERVICE_PREFIX,`

[the remark this answers is not in the document]
Comment (umustafi) [reply gobblin-metrics-libs/gobblin-metrics/src/main/java/org/apache/gobblin/metrics/ServiceMetricNames.java:38]: Additionally, I don't want to replace this prefix in configuration keys right now as I'll have to go and change the config everywhere and may run into some issues.

**Comment by umustafi** [comment]
Tried to add a unit test for `DagActionStoreMonitor.processMessage` to ensure that `heartbeat` messages or null `dagActionType` messages are processed in the correct order but these tests will need to be e2e tests that instantiate a `DagActionStoreMonitor` instance, created DecodeableKafkaRecords, and can properly parse a variety of messages (insert, delete, heartbeat, etc...) but it's going to be a bigger change and LOE than the scope of the current PR. I will follow up with a separate PR for that.

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/ChangeMonitorUtils.java [hunk 3]
```diff
@@ -28,20 +30,24 @@ private ChangeMonitorUtils() {
...
-   * the pre-conditions above don't apply and we should proceed processing the change event
+   * Performs checks for duplicate messages, heartbeat message types, or null dag action types all of which cannot or
+   * should not be processed. Returns true if the pre-conditions above don't apply, and we should proceed processing
+   * the change event
    */
-  public static boolean shouldProcessMessage(String changeIdentifier, LoadingCache<String, String> cache,
-      String operation, String timestamp) {
+  public static boolean isValidAndUniqueMessage(String changeIdentifier, LoadingCache<String, String> cache,
+      String operation, String timestamp, ContextAwareMeter duplicateMessagesMeter,
+      ContextAwareMeter heartbeatMessagesMeter) {
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/ChangeMonitorUtils.java:39]: nit: I would put all the elements from the message together as the first params, and then the cache and the meters

**Review by phet — APPROVED** [review]
looks good!

**Review by phet — APPROVED** [review]

**Review by Will-Lo — APPROVED** [review]

**Pull request merged**